### PR TITLE
Light Client: Evidence reporting

### LINKED ISSUE #217: [light-client] Evidence Reporting
The light client is responsible for submitting evidence of malfeasance outlined in [Tendermint-go ADR-047](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-047-handling-evidence-from-light-client.md). The rust implementation will also need to detect and submit evidence to connected full nodes.

### PR DESCRIPTION
**Pull request by romac**
Closes #217

## TODO

- [x] Merge `master` in
- [ ] Documentation
- [ ] Tests
- [x] Match the spec more closely (see #349)
- [ ] Add tags to the code for traceability
- [ ] Add contracts wherever possible

---

* [x] Referenced an issue explaining the need for the change
* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGES.md

### CODE REVIEW

File: light-client/src/fork_detector.rs [hunk 1]
```diff
@@ -109,8 +112,16 @@ impl ForkDetector for ProdForkDetector {
                 .verify_to_target(light_block.height(), &mut state);
 
             match result {
-                Ok(_) => forks.push(Fork::Forked(witness_block)),
-                Err(e) if e.kind().has_expired() => forks.push(Fork::Forked(witness_block)),
+                Ok(_) => forks.push(Fork::Forked {
+                    primary: light_block.clone(),
+                    witness: witness_block,
+                }),
+                Err(e) if e.kind().has_expired() => {
```
Comment (melekes) [inline light-client/src/fork_detector.rs:119]:
```suggestion
                Err(e) if !e.kind().has_expired() => {
```

?

Comment (romac) [reply light-client/src/fork_detector.rs:119]: Could you please expand on that?

The spec currently describe the case where header expired as a potential fork: https://github.com/informalsystems/tendermint-rs/pull/287/files#diff-57dd6c1989dd85c55b47487b7e1f92d8R409-R414

**Comment by romac** [comment]
> what's left here? writing tests?

- Documentation
- Tests
- Match the spec more closely (eg. maintain a list of full nodes that haven't been contacted)
- Add tags to the code for traceability
- Add contracts wherever possible

**Comment by brapse** [comment]
Could this be exercised by #331 perhaps?

**Comment by romac** [comment]
> Could this be exercised by #331 perhaps?

You mean the whole fork detector? I guess that would exercise part of it, yes but not the case where the header verifies successfully.

File: light-client/src/fork_detector.rs (hunk 1 above)
Comment (melekes) [reply light-client/src/fork_detector.rs:119]: Right. We report it even though we won't be able to punish. All good

File: tendermint/src/evidence.rs [hunk 2]
```diff
@@ -30,6 +34,22 @@ pub struct DuplicateVoteEvidence {
...
+    #[serde(rename = "H2")]
+    h2: SignedHeader,
+}
+
+impl ConflictingHeadersEvidence {
+    /// Create a new evidence of conflicting headers
+    pub fn new(h1: SignedHeader, h2: SignedHeader) -> Self {
+        Self { h1, h2 }
+    }
+}
```
Comment (Liamsi) [inline tendermint/src/evidence.rs:51]: same

**Pull request merged**